assets/js/app.js

```diff
@@ -0,0 +1,431 @@
+                console.warn("No data received for cryptos");
+                return;
+            }
+
+            coins.forEach((coin, index) => {
+                const cap = coin.toUpperCase();
+                if (graphData[cap] && graphData[cap].USD !== undefined) {
+                    const price = graphData[cap].USD;
+                    data.labels.push(currentTime);
+                    data.datasets[index].data.push({ x: currentTime, y: price });
+                } else {
+                    console.warn(`The price for ${cap} could not be retrieved.`);
+                    data.labels.push(currentTime);
+                    data.datasets[index].data.push({ x: currentTime, y: 0 });
+                }
+
+                if (data.labels.length > 20) {
+                    data.labels.shift();
+                    data.datasets[index].data.shift();
+                }
+            });
+
+            chart.update();
+        };
+
+        updateGraph();
+
+        updateInterval = setInterval(updateGraph, 2000);
+    };
+
+    const getRandomColor = () => {
+        const letters = '0123456789ABCDEF';
+        let color = '#';
+        for (let i = 0; i < 6; i++) {
+            color += letters[Math.floor(Math.random() * 16)];
+        }
+        return color;
+    };
+
+    const resetGraph = () => {
+        if (chart) {
+            chart.destroy();
+            chart = null;
+        }
+
+        if (updateInterval) {
+            clearInterval(updateInterval);
+            updateInterval = null;
+        }
+    };
+
+    const renderCoinsHTML = newHTML => document.getElementById("cardsContainer").innerHTML = newHTML ? newHTML : `
+        <div class="alert alert-warning alert-dismissible fade show" role="alert">
+            <strong>Warning!</strong> No results found
+        </div>
+    `;
+
+    const addEventButton = () => {
+        const popoverTriggerList = document.querySelectorAll('[data-bs-toggle="popover"]');
+
+        popoverTriggerList.forEach(popoverTriggerEl => {
+            const popover = new bootstrap.Popover(popoverTriggerEl, {
+                trigger: "manual",
+                html: true
+            });
+
+            popoverTriggerEl.addEventListener('focusout', () => {
+                popover.hide();
+            });
+
+            popoverTriggerEl.addEventListener("click", async function () {
+                const progressBarContainer = document.getElementById(`progressBarContainer-${this.id}`);
+                const progressBar = document.getElementById(`progressBar-${this.id}`);
+
+                progressBarContainer.style.display = 'block';
+                progressBar.style.width = '0%';
+                progressBar.textContent = '0%';
+
+                const coin = await getSingleCoin(this.id);
+
+                let progress = 0;
+                const interval = setInterval(() => {
+                    if (progress < 100) {
+                        progress += 5;
```

Comment:
this turns out weird when you're actually fetching the data from the cache. no need for pyrotechnics if the user is held back from the info they need